hubgrep/frontend_blueprint/forms/hosting_service/hosting_services.py

```diff
@@ -0,0 +1,164 @@
+import json
+import logging
+
+import markdown
+
+from flask_wtf import FlaskForm
+from wtforms import StringField, SelectField, TextAreaField
+from wtforms.widgets.html5 import URLInput
+from wtforms.validators import ValidationError, URL
+from wtforms.validators import InputRequired
+
+from hubgrep.lib.hosting_service_interfaces import hosting_service_interface_mapping
+from hubgrep.models import HostingService
+
+from hubgrep.frontend_blueprint.forms.hosting_service.validators import (
+    validate_custom_config,
+    validate_url,
+    validate_api_url,
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+class NoHostingServiceFormException(Exception):
+    pass
+
+
+class HostingServiceFormFirstStep(FlaskForm):
+    """ Only get type and landingpage from the user, the rest we automatically resolve to pre-fill in step 2. """
+
+    type = SelectField(
+        u"Type",
+        [InputRequired()],
+        choices=[
+            ("gitea", "gitea"),
+            ("gitlab", "Gitlab"),
+            ("github", "GitHub"),
+        ],
+    )
+    landingpage_url = StringField(
+        "Landingpage Url",
+        [InputRequired(), URL(), validate_url],
+        widget=URLInput(),
+        description="the landingpage for this hoster"
+    )
+
+
+class HostingServiceForm(HostingServiceFormFirstStep):
+    """ Final step when adding a new hosting-service, doubling as the form used for editing hosting-services. """
+
+    # validate_url sanitizes url before validate_api_url makes a call to it
+    api_url = StringField(
+        "Api Url",
+        [InputRequired(), URL(), validate_url, validate_api_url],
```

Comment:
validate_url "sanitizes" the user input, validate_api_url makes an actual hostingserviceinterface instance and checks if the output looks like what we expect (see HostingServiceInterface classes)

Reply:
sounds good